contracts/src/contracts/ValidatorManager/BalancerValidatorManager.sol

```diff
@@ -0,0 +1,320 @@
+            )
+        );
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function getChurnPeriodSeconds() external view returns (uint64 churnPeriodSeconds) {
+        return _getChurnPeriodSeconds();
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function getSecurityModules() external view returns (address[] memory securityModules) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        return $.securityModules.keys();
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function getSecurityModuleWeights(
+        address securityModule
+    ) external view returns (uint64 weight, uint64 maxWeight) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        weight = $.securityModuleWeight[securityModule];
+        maxWeight = uint64($.securityModules.get(securityModule));
+
+        return (weight, maxWeight);
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function isValidatorPendingWeightUpdate(
+        bytes32 validationID
+    ) external view returns (bool) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        return $.validatorPendingWeightUpdate[validationID] != 0;
+    }
+
+    function _checkValidatorSecurityModule(
+        bytes32 validationID,
+        address securityModule
+    ) internal view {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        // If the validator has no associated security module, it is an initial validator
+        // and can be managed by any security module
+        if ($.validatorSecurityModule[validationID] == address(0)) {
+            return;
+        } else if ($.validatorSecurityModule[validationID] != securityModule) {
+            revert BalancerValidatorManager__ValidatorNotBelongingToSecurityModule(
+                validationID, securityModule
+            );
+        }
+    }
+
+    function _setupSecurityModule(address securityModule, uint64 maxWeight) internal {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        if (maxWeight == 0) {
+            if (!$.securityModules.remove(securityModule)) {
+                revert BalancerValidatorManager__SecurityModuleNotRegistered(securityModule);
+            }
+        } else {
+            $.securityModules.set(securityModule, uint256(maxWeight));
+        }
+
+        emit SetupSecurityModule(securityModule, maxWeight);
+    }
+
+    function _updateSecurityModuleWeight(address securityModule, uint64 weight) internal {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        if (weight > uint64($.securityModules.get(securityModule))) {
+            revert BalancerValidatorManager__SecurityModuleMaxWeightReached(securityModule, weight);
+        }
+
+        $.securityModuleWeight[securityModule] = weight;
```

Comment:
Shouldn't we increment/decrement `$.l1TotalWeight` by `$.securityModuleWeight[securityModule] - weight` here?

Reply:
Yes, `$.l1TotalWeight` should be updated as well, but it is actually very hard to do without an access to `VALIDATOR_MANAGER_STORAGE_LOCATION`.